include/openssl/ec.h

```diff
@@ -1228,7 +1228,7 @@ int ECDSA_sign_setup(EC_KEY *eckey, BN_CTX *ctx, BIGNUM **kinv, BIGNUM **rp);
 
 /** Computes ECDSA signature of a given hash value using the supplied
  *  private key (note: sig must point to ECDSA_size(eckey) bytes of memory).
- *  \param  type     this parameter is ignored
+ *  \param  type     non-positive value will be ignored, otherwise, hash nid for RFC6979 deterministic signature generation
```

Comment:
Thinking a bit more about this - it will need a new API function. The existing API is public which means you cant assume that type will be -1 in everyone existing code.
So do not change this function - make a new function for the deterministic case.. And then have both functions just call an internal function that does the work for both cases.
Note also that there is a nid for NID_undef (which is 0).

Reply:
@slontis 
> make a new function for the deterministic case.. And then have both functions just call an internal function that does the work for both cases.

How about call eckey->meth->sign directly in EVP (pctx->meth->sign) ?
Is a new function really needed?

Reply:
Can you use ctrls to set whatever you need to pass in (the same way you set the nonce_type)?

Reply:
@slontis I don't really understand what you means...
There isn't a ctrl method in eckey->meth
Are you implying that set nid into eckey->ex_data?

Reply:
I was thinking the same way you have called EVP_PKEY_CTX_set_ecdsa_nonce_type() inside pkey_ec_ctrl(). Couldnt you just set the nid the same way? (I could be missing something)

Reply:
I still don't understand...
pkey_ec_ctrl() sets the hash nid into pctx->data, namely dctx, but neither pctx nor dctx will be passed to the call to ECDSA_sign to ECDSA_sign_ex to eckey->sign (ossl_ecdsa_sign) to ecdsa_sign_setup
how would you like the nid to be known to the ecdsa_sign_setup?

Reply:
I dont know the answer. The only way to pass it in is to either
1) Have a seperate API.
2) Put it inside EC_KEY (somehow?)
@levitte any ideas on how to plumb this in?

Reply:
@slontis I mean that EVP_PKEY _CTX pctx and EC_PKEY_CTX dctx won't be passed to ecdsa_sign_setup, which initialize k_inverse and r. So the nid has to be read from dctx, and pass to ecdsa_sign_setup. 
I want to skip ECDSA_sign and ECDSA _sign_ex, as I can't change their behavior, so I can't pass the nid to them.
If these two methods are skipped, then pctx->meth->sign would call eckey->meth->sign directly.